students/JeroenSweerts/Lesson5/mailroom3.py

```diff
@@ -0,0 +1,105 @@
+
+
+def createfile(name):
+    '''create .txt file containing thank you letter for each person who made a donation'''
+    d = {'name':name, 'donation':donations[name][-1]}
+    outfile = open(name+'.txt', 'w')
+    outfile.write("Dear {name} , \n\n".format(**d))
+    outfile.write("\t Thank you very much for your generous donation of ${donation}. \n\n".format(**d))
+    outfile.write("\t It will be put to very good use. \n\n")
+    outfile.write("\t\t Sincerely, \n")
+    outfile.write("\t\t   -The Team \n")
+    outfile.close()
+
+
+def thankyoueveryone():
+    [createfile(name) for name in donations.keys()]
+
+
+def report():
+    """ If the user (you) selected “Create a Report”, print a list of your donors,
+     sorted by total historical donation amount.
+     Include Donor Name, total donated, number of donations and average donation
+     amount as values in each row. You do not need to print out all their
+     donations, just the summary info.
+     Using string formatting, format the output rows as nicely as possible.
+     The end result should be tabular (values in each column should align with
+     those above and below)
+     After printing this report, return to the original prompt.
+     At any point, the user should be able to quit their current task and return
+     to the original prompt.
+     From the original prompt, the user should be able to quit the script cleanly."""
+    print(f'{"Donor Name":20s} {"|  Total Given":20s} {"|  Num Gifts  |":20s} {"Average Gift":20s}')
+    print(f'{"-"*76}')
+    [print(f'{name:20s} ${sum(donations[name]):20.2f} {len(donations[name]):13d}${sum(donations[name])/len(donations[name]):20.2f}') for name in donations.keys()]
+
+
+if __name__ == '__main__':
+    response = input('Please choose between the following 3 actions:\
+    1. Send a Thank You; 2. Create a Report; 3. Send letters to everyone; 4. quit >  ')
+    try:
+        int(response)
+    except:
+        while not response.isdigit():
+            print("You didn't enter a number.")
+            print()
+            response = input('Please choose between the following 3 actions:\
+            1. Send a Thank You; 2. Create a Report; 3. Send letters to everyone; 4. quit >  ')
+    arg_dict = {1:thankyou, 2:report, 3:thankyoueveryone}
+    while int(response) != 4:
+        try:
+            arg_dict.get(int(response))()
+        except (TypeError, ValueError):
+            print('This is an invalid choice. Please enter a number between 1 and 4.')
+            print()
+            response = input('Please choose between the following 3 actions:\
+            1. Send a Thank You; 2. Create a Report; 3. Send letters to everyone; 4. quit >  ')
+        else:
+            response = input('Please choose between the following 3 actions:\
+                1. Send a Thank You; 2. Create a Report; 3. Send letters to everyone; 4. quit >   ')
+        finally:
+            if not response.isdigit():
+                response = '99999'
```

Comment:
this is a good practice here for try/except but sometimes keeping things simple is best 😉 

consider this alternative:
```
options = {'1': thank_you, '2': report, '3': thank_you_everyone}
while True:
    option = input("Please choose.... ")
    try:
        options[option]()
    except KeyError:
        print('valid options are: %s' % ', '.join(list(options)): 
```
above should take care of pretty much all of the logic